flocker/node/functional/test_docker.py

```diff
@@ -635,32 +636,20 @@ def run_registry(self):
                     internal_port=5000,
                     # Doesn't matter what port we expose this on.  We'll
                     # discover what was assigned later.
-                    external_port=0,
+                    external_port=external_port,
                 ),
             ],
             retry_on_port_collision=True,
         )
 
-        def extract_listening_port(client):
-            listing = client.list()
-            listing.addCallback(
-                lambda applications: list(
-                    next(iter(application.ports)).external_port
-                    for application
-                    in applications
-                )[0]
```

Comment:
If your hypothesis is correct (and it certainly seems plausible), then maybe a good first step is to raise an error here if there is more than one application. Although it's not a final solution, it will be an incremental improvement over a timeout (faster failure; more informative error message; better logs).